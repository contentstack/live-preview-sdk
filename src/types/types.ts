import {
    IInviteMetadata,
    IThreadDTO,
} from "../visualBuilder/types/collab.types";
export declare interface IEditEntrySearchParams {
    hash?: string;
    entry_uid?: string;
    content_type_uid?: string;
    /**
     * @deprecated pass this value as hash instead
     */
    live_preview?: string;
}

export declare interface IClientUrlParams {
    protocol: "http" | "https";
    host: string;
    port: string | number;
    url: string;
}

export declare interface IStackSdk {
    live_preview: { [key: string]: any } & Partial<IConfig>;
    [key: string]: any;
    environment: string;
}

export declare interface IStackDetails {
    apiKey: string;
    environment: string;
    contentTypeUid: string;
    entryUid: string;
    branch: string;
    /**
     * This locale is currently used by the visual builder to
     * redirect to the correct locale if the no data-cslp tag
     * is present in the HTML to extract the locale.
     */
    locale: string;
    masterLocale: string;
}

export declare interface IInitStackDetails {
    apiKey: string;
    environment: string;
    branch: string;
    /**
     * This locale is currently used by the visual builder to
     * redirect to the correct locale if the no data-cslp tag
     * is present in the HTML to extract the locale.
     */
    locale: string;
}

export declare type ILivePreviewMode = "builder" | "preview";

//? We kept it as number so that we could disable only the unrequired features,
//? since the "Builder" mode will contain all the features of the "Preview" mode.
export enum ILivePreviewModeConfig {
    PREVIEW = 1,
    BUILDER = 2,
}

export enum ILivePreviewWindowType {
    PREVIEW = "preview",
    PREVIEW_SHARE = "preview-share",
    BUILDER = "builder",
    INDEPENDENT = "independent",
}

export declare interface IConfig {
    ssr: boolean;
    enable: boolean;
    /**
     * @default false
     */
    debug: boolean;
    cleanCslpOnProduction: boolean;
    stackDetails: IStackDetails;
    clientUrlParams: IClientUrlParams;
    stackSdk: IStackSdk;
    onChange: () => void;
    runScriptsOnUpdate: boolean;
    windowType: ILivePreviewWindowType;
    hash: string;
    editButton: IConfigEditButton;
    mode: ILivePreviewModeConfig;
    elements: {
        highlightedElement: HTMLElement | null;
    };
    collab: {
        enable: boolean;
        state: boolean;
        inviteMetadata: IInviteMetadata;
    };
}

export declare interface IConfigEditButton {
    enable: boolean;
    exclude?: ("insideLivePreviewPortal" | "outsideLivePreviewPortal")[];
    includeByQueryParameter?: boolean;
    position?:
        | "top"
        | "bottom"
        | "left"
        | "right"
        | "top-left"
        | "top-right"
        | "top-center"
        | "bottom-left"
        | "bottom-right"
        | "bottom-center";
}

export declare interface IInitData {
    ssr: boolean;
    runScriptsOnUpdate: boolean;
    enable: boolean;
    /**
     * @default false
     */
    debug: boolean;
    cleanCslpOnProduction: boolean;
    stackDetails: Partial<IInitStackDetails>;
    clientUrlParams: Partial<Omit<IClientUrlParams, "url">>;
    stackSdk: IStackSdk;
    editButton: IConfigEditButton;
    mode: ILivePreviewMode;
}

// type PickPartial<T, K extends keyof T> = Partial<Pick<T, K>> & Omit<T, K>;

// export type IInitData =
//     | {
//           ssr?: true;
//           stackSdk?: IStackSdk;
//           runScriptsOnUpdate?: boolean;
//           enable?: boolean;
//           cleanCslpOnProduction?: boolean;
//           stackDetails?: Partial<IInitStackDetails>;
//           clientUrlParams?: Partial<Omit<IClientUrlParams, "url">>;
//           editButton?: IConfigEditButton;
//           mode?: "preview";
//       }
//     | {
//           ssr?: true;
//           stackSdk?: IStackSdk;
//           runScriptsOnUpdate?: boolean;
//           enable?: boolean;
//           cleanCslpOnProduction?: boolean;
//           stackDetails: PickPartial<IInitStackDetails, "branch">;
//           clientUrlParams: PickPartial<
//               Omit<IClientUrlParams, "url">,
//               "port" | "protocol"
//           >;
//           editButton?: IConfigEditButton;
//           mode: "builder";
//       }
//     | {
//           ssr: false;
//           runScriptsOnUpdate?: boolean;
//           enable?: boolean;
//           cleanCslpOnProduction?: boolean;
//           stackDetails?: Partial<IInitStackDetails>;
//           clientUrlParams?: Partial<Omit<IClientUrlParams, "url">>;
//           stackSdk: IStackSdk;
//           editButton?: IConfigEditButton;
//           mode?: "preview";
//       }
//     | {
//           ssr: false;
//           runScriptsOnUpdate?: boolean;
//           enable?: boolean;
//           cleanCslpOnProduction?: boolean;
//           stackDetails: PickPartial<IInitStackDetails, "branch">;
//           clientUrlParams: PickPartial<
//               Omit<IClientUrlParams, "url">,
//               "port" | "protocol"
//           >;
//           stackSdk: IStackSdk;
//           editButton?: IConfigEditButton;
//           mode: "builder";
//       };

// Post message types

export declare interface ILivePreviewMessageCommon {
    from: "live-preview";
}

export declare interface IEditButtonPosition {
    upperBoundOfTooltip: number;
    leftBoundOfTooltip: number;
}

// end of Post message types

export interface IVisualBuilderInitEvent {
    windowType: ILivePreviewWindowType;
    stackDetails: {
        masterLocale: string;
    };
    collab?: {
        enable: boolean;
        state: boolean;
        inviteMetadata: IInviteMetadata;
        payload: IThreadDTO[];
    };
}
