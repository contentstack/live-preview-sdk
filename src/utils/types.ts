export declare interface IEditEntrySearchParams {
    hash?: string;
    entry_uid?: string;
    content_type_uid?: string;
    /**
     * @deprecated pass this value as hash instead
     */
    live_preview?: string;
}

export declare interface IClientUrlParams {
    protocol: "http" | "https";
    host: string;
    port: string | number;
    url: string;
}

export declare interface IStackSdk {
    live_preview: { [key: string]: any } & Partial<IConfig>;
    [key: string]: any;
    headers: {
        api_key: string;
    };
    environment: string;
}

export declare interface IStackDetails {
    apiKey: string;
    environment: string;
    contentTypeUid: string;
    entryUid: string;
}

export declare interface IInitStackDetails {
    apiKey: string;
    environment: string;
}
export declare interface IConfig {
    ssr: boolean;
    enable: boolean;
    cleanCslpOnProduction: boolean;
    stackDetails: IStackDetails;
    clientUrlParams: IClientUrlParams;
    stackSdk: IStackSdk;
    onChange: () => void;
    runScriptsOnUpdate: boolean;
}

export declare interface IInitData {
    ssr: boolean;
    runScriptsOnUpdate: boolean;
    enable: boolean;
    cleanCslpOnProduction: boolean;
    stackDetails: Partial<IInitStackDetails>;
    clientUrlParams: Partial<Omit<IClientUrlParams, "url">>;
    stackSdk: IStackSdk;
}

// Post message types

export declare interface ILivePreviewMessageCommon {
    from: "live-preview";
}

export type ILivePreviewReceivePostMessages =
    | IClientDataMessage
    | IInitAckMessage
    | IHistoryMessage
    | IDocWithScriptMessage;
export declare interface IClientDataMessage extends ILivePreviewMessageCommon {
    type: "client-data-send";
    data: {
        hash: string;
    };
}

export declare interface IInitAckMessage extends ILivePreviewMessageCommon {
    type: "init-ack";
    data: {
        contentTypeUid: string;
        entryUid: string;
    };
}

export declare interface IHistoryMessage extends ILivePreviewMessageCommon {
    type: "history";
    data: {
        type: "forward" | "backward" | "reload";
    };
}

export declare interface IDocWithScriptMessage
    extends ILivePreviewMessageCommon {
    type: "document-body-post-scripts-loaded";
    data: {
        body: string;
    };
}

// end of Post message types

export declare type OnEntryChangeCallback = () => void;

export declare type OnEntryChangeCallbackUID = string;
