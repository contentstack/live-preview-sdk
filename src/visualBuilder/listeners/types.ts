import { VisualEditorCslpEventDetails } from "../types/visualBuilder.types";

interface EventListenerHandlerParams {
    event: MouseEvent;
    overlayWrapper: HTMLDivElement | null;
    visualEditorContainer: HTMLDivElement | null;
    previousSelectedEditableDOM: Element | null;
    focusedToolbar: HTMLDivElement | null;
    resizeObserver: ResizeObserver;
    eventDetails: VisualEditorCslpEventDetails;
    customCursor: HTMLDivElement | null;
}

export default EventListenerHandlerParams;
