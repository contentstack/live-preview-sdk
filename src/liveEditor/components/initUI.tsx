import { h, render } from "preact";
import VisualEditorComponent from "./visualEditor";

interface InitUIParams {
    previousSelectedEditableDOM: HTMLDivElement | null;
    resizeObserver: ResizeObserver;
}

function initUI(props: InitUIParams) : void {

    console.log('[IN SDK] : initUI : creating root node');

    const visualEditorDOM = document.querySelector(
        ".visual-editor__container"
    );
    if (!visualEditorDOM) {
       
        const visualEditorWrapper = document.createElement("div");
        visualEditorWrapper.classList.add("visual-editor__container");
        visualEditorWrapper.setAttribute("data-testid", "visual-editor__container");

        document.body.appendChild(visualEditorWrapper);
        
        render(<VisualEditorComponent 
            visualEditorWrapper={visualEditorWrapper}
            previousSelectedEditableDOM={props.previousSelectedEditableDOM}
            resizeObserver={props.resizeObserver}
        />, visualEditorWrapper);
        
        console.log('[IN SDK] : initUI : rendered root node');
    }
    
    return;

}

export default initUI;
