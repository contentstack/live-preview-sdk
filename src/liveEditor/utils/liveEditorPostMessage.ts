import { EventManager } from "@contentstack/advanced-post-message";
import { LIVE_EDITOR_CHANNEL_ID } from "./constants";

let liveEditorPostMessage: EventManager | undefined;

if (typeof window !== "undefined") {
    liveEditorPostMessage = new EventManager(LIVE_EDITOR_CHANNEL_ID, {
        target: window.parent,
        debug: false,
    });
}

export default liveEditorPostMessage;
