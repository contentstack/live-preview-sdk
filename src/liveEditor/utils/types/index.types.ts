export interface ITraverseSchemaVisitor extends Record<string, any> {
    should_visit: (fieldSchema: any, path: string) => boolean;
    visit: (fieldSchema: any, path: string) => void;
}
