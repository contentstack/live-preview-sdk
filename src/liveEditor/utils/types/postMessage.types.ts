export enum LiveEditorPostMessageEvents {
    ADD_INSTANCE = "add-instance",
    UPDATE_FIELD = "update-field",
    OPEN_ASSET_MODAL = "open-asset-modal",
    OPEN_QUICK_FORM = "open-quick-form",
    GET_FIELD_SCHEMA = "get-field-schema",
    FOCUS_FIELD = "focus-field",
    DELETE_INSTANCE = "delete-instance",
    MOVE_INSTANCE = "move-instance",
}
