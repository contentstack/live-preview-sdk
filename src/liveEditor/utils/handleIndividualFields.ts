import { VisualEditorCslpEventDetails } from "../../types/liveEditor.types";
import {
    generateReplaceAssetButton,
    removeReplaceAssetButton,
} from "./assetButton";
import { LIVE_EDITOR_FIELD_TYPE_ATTRIBUTE_KEY } from "./constants";
import { getFieldType } from "./getFieldType";
import { handleFieldInput, handleFieldKeyDown } from "./handleFieldMouseDown";
import liveEditorPostMessage from "./liveEditorPostMessage";
import {
    handleAddButtonsForMultiple,
    removeAddInstanceButtons,
} from "./multipleElementAddButton";
import { FieldDataType } from "./types/index.types";
import { LiveEditorPostMessageEvents } from "./types/postMessage.types";

/**
 * It handles all the fields based on their data type and its "multiple" property.
 * @param eventDetails The event details object that contain cslp and field metadata.
 * @param elements The elements object that contain the visual editor wrapper.
 */
export function handleIndividualFields(
    eventDetails: VisualEditorCslpEventDetails,
    elements: {
        visualEditorWrapper: HTMLDivElement;
        lastEditedField: Element | null;
    }
): void {
    const { fieldSchema, editableElement } = eventDetails;
    const { visualEditorWrapper, lastEditedField } = elements;
    const fieldType = getFieldType(fieldSchema);

    editableElement.setAttribute(
        LIVE_EDITOR_FIELD_TYPE_ATTRIBUTE_KEY,
        fieldType
    );

    if (
        // @ts-ignore
        fieldSchema?.multiple ||
        (fieldSchema.data_type === "reference" &&
            // @ts-ignore
            fieldSchema.field_metadata.ref_multiple)
    ) {
        if (lastEditedField !== editableElement) {
            handleAddButtonsForMultiple(eventDetails, {
                editableElement: eventDetails.editableElement,
                visualEditorWrapper: visualEditorWrapper,
            });
        }

        // * fields could be handled as they are in a single instance
        if (eventDetails.fieldMetadata.multipleFieldMetadata.index > -1) {
            handleSingleField();
        }
        return;
    } else {
        handleSingleField();
    }

    /**
     * Handles all the fields based on their data type.
     */
    function handleSingleField() {
        /**
         * The field that can be directly modified using contenteditable=true.
         * This includes all text fields like title and numbers.
         */
        const ALLOWED_INLINE_EDITABLE_FIELD: FieldDataType[] = [
            FieldDataType.SINGLELINE,
            FieldDataType.MULTILINE,
            FieldDataType.NUMBER,
        ];

        // * title, single single_line, single multi_line, single number
        if (ALLOWED_INLINE_EDITABLE_FIELD.includes(fieldType)) {
            editableElement.setAttribute("contenteditable", "true");
            editableElement.addEventListener("input", handleFieldInput);
            editableElement.addEventListener("keydown", handleFieldKeyDown);

            return;
        }

        if (fieldSchema.data_type === "file") {
            const replaceButton = generateReplaceAssetButton(
                editableElement,
                () => {
                    liveEditorPostMessage?.send(
                        LiveEditorPostMessageEvents.OPEN_ASSET_MODAL,
                        {
                            fieldMetadata: eventDetails.fieldMetadata,
                        }
                    );
                }
            );

            visualEditorWrapper?.appendChild(replaceButton);
            return;
        }

        liveEditorPostMessage?.send(
            LiveEditorPostMessageEvents.OPEN_QUICK_FORM,
            {
                fieldMetadata: eventDetails.fieldMetadata,
                cslpData: eventDetails.cslpData,
            }
        );
    }
}

export function cleanIndividualFieldResidual(elements: {
    overlayWrapper: HTMLDivElement;
    previousSelectedEditableDOM: Element;
    visualEditorWrapper: HTMLDivElement | null;
}): void {
    const { overlayWrapper, previousSelectedEditableDOM, visualEditorWrapper } =
        elements;

    removeAddInstanceButtons({
        eventTarget: null,
        visualEditorWrapper: visualEditorWrapper,
        overlayWrapper: overlayWrapper,
    });

    removeReplaceAssetButton(visualEditorWrapper);

    previousSelectedEditableDOM.removeAttribute(
        LIVE_EDITOR_FIELD_TYPE_ATTRIBUTE_KEY
    );
    previousSelectedEditableDOM.removeAttribute("contenteditable");
    previousSelectedEditableDOM.removeEventListener("input", handleFieldInput);
    previousSelectedEditableDOM.removeEventListener(
        "keydown",
        handleFieldKeyDown
    );
}
