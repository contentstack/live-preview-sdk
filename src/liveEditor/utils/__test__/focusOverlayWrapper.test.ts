import "@testing-library/jest-dom/extend-expect";
import { fireEvent } from "@testing-library/preact";
import {
    addFocusOverlay,
    hideFocusOverlay,
} from "./../../generators/generateOverlay";
import initUI from "../../components";
import { cleanIndividualFieldResidual } from "../handleIndividualFields";
import { LiveEditorPostMessageEvents } from "../types/postMessage.types";
import liveEditorPostMessage from "../liveEditorPostMessage";
import { VisualEditor } from "../..";

jest.mock("../liveEditorPostMessage", () => {
    return {
        __esModule: true,
        default: {
            send: jest.fn(),
            on: jest.fn(),
        },
    };
});

jest.mock("../handleIndividualFields", () => {
    return {
        __esModule: true,
        cleanIndividualFieldResidual: jest.fn(),
    };
});

const mockResizeObserver = {
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
};

describe("addFocusOverlay", () => {
    let visualEditorContainer: HTMLDivElement;
    let targetElement: HTMLElement;
    let focusOverlayWrapper: HTMLDivElement;

    jest.mock("../../generators/generateOverlay", () => {
        const originalModule = jest.requireActual(
            "../../generators/generateOverlay"
        );
        return {
            ...originalModule,
            hideOverlay: jest.fn(),
        };
    });

    beforeEach(() => {
        initUI({
            resizeObserver: mockResizeObserver,
        });
        visualEditorContainer = document.querySelector(
            ".visual-editor__container"
        ) as HTMLDivElement;

        focusOverlayWrapper = document.querySelector(
            ".visual-editor__overlay__wrapper"
        ) as HTMLDivElement;

        // Create a target element and focus overlay wrapper for each test
        targetElement = document.createElement("p");
        targetElement.setAttribute(
            "data-cslp",
            "all_fields.blt58a50b4cebae75c5.en-us.title"
        );
        targetElement.getBoundingClientRect = jest.fn(() => ({
            left: 10,
            right: 20,
            top: 10,
            bottom: 20,
        })) as any;

        visualEditorContainer.appendChild(targetElement);
    });

    afterEach(() => {
        // Remove the target element and focus overlay wrapper after each test
        document.body.innerHTML = "";
        jest.clearAllMocks();
    });

    test("should set the focus overlay wrapper to visible", () => {
        addFocusOverlay(targetElement, focusOverlayWrapper);
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);
    });

    test("should set the top overlay to the correct dimensions", () => {
        expect(focusOverlayWrapper).toMatchSnapshot();

        addFocusOverlay(targetElement, focusOverlayWrapper);

        focusOverlayWrapper = document.querySelector(
            ".visual-editor__overlay__wrapper"
        ) as HTMLDivElement;

        expect(focusOverlayWrapper).toMatchSnapshot();
        expect(focusOverlayWrapper?.classList.contains("visible")).toBe(true);

        const visualEditorWrapperTopOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--top"]`
        ) as HTMLDivElement;
        const visualEditorWrapperLeftOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--left"]`
        ) as HTMLDivElement;
        const visualEditorWrapperRightOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--right"]`
        ) as HTMLDivElement;
        const visualEditorWrapperBottomOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--bottom"]`
        ) as HTMLDivElement;

        expect(visualEditorWrapperTopOverlay.style.top).toBe("0px");
        expect(visualEditorWrapperTopOverlay.style.left).toBe("0px");
        expect(visualEditorWrapperTopOverlay.style.width).toBe("100%");
        expect(visualEditorWrapperTopOverlay.style.height).toBe(
            "calc(10px - 2px)"
        );

        expect(visualEditorWrapperBottomOverlay.style.top).toBe(
            "calc(20px + 2px)"
        );
        expect(visualEditorWrapperBottomOverlay.style.left).toBe("0px");
        expect(visualEditorWrapperBottomOverlay.style.width).toBe("100%");
        expect(visualEditorWrapperBottomOverlay.style.height).toBe(
            "calc(-20px - 2px)"
        );

        expect(visualEditorWrapperLeftOverlay.style.top).toBe(
            "calc(10px - 2px)"
        );
        expect(visualEditorWrapperLeftOverlay.style.left).toBe("0px");
        expect(visualEditorWrapperLeftOverlay.style.width).toBe(
            "calc(10px - 2px)"
        );

        expect(visualEditorWrapperRightOverlay.style.top).toBe(
            "calc(10px - 2px)"
        );
        expect(visualEditorWrapperRightOverlay.style.left).toBe(
            "calc(20px + 2px)"
        );
        expect(visualEditorWrapperRightOverlay.style.width).toBe(
            "calc(1004px - 2px)"
        );
    });
});

describe("hideFocusOverlay", () => {
    let visualEditorContainer: HTMLDivElement;
    let editedElement: HTMLParagraphElement;
    let focusOverlayWrapper: HTMLDivElement;
    let singleFocusOverlay: HTMLDivElement;

    beforeEach(() => {
        initUI({
            resizeObserver: mockResizeObserver,
        });
        visualEditorContainer = document.querySelector(
            ".visual-editor__container"
        ) as HTMLDivElement;

        focusOverlayWrapper = document.querySelector(
            ".visual-editor__overlay__wrapper"
        ) as HTMLDivElement;

        editedElement = document.createElement("p");
        editedElement.setAttribute(
            "data-cslp",
            "all_fields.blt58a50b4cebae75c5.en-us.title"
        );
        editedElement.innerHTML = "Hello World!";
        editedElement.getBoundingClientRect = jest.fn(() => ({
            left: 10,
            right: 20,
            top: 10,
            bottom: 20,
        })) as any;

        document.body.appendChild(editedElement);

        singleFocusOverlay = focusOverlayWrapper.querySelector(
            `[data-testid="visual-editor__overlay--top"]`
        ) as HTMLDivElement;

        addFocusOverlay(editedElement, focusOverlayWrapper);
    });

    afterEach(() => {
        document.body.innerHTML = "";
        jest.clearAllMocks();
    });

    test("should not hide the overlay if the focus overlay wrapper is null", () => {
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);

        hideFocusOverlay({
            visualEditorContainer,
            visualEditorOverlayWrapper: null,
            focusedToolbar: null,
        });

        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);
    });

    test("should hide the overlay if the focus overlay is clicked", () => {
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);
        singleFocusOverlay.click();
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(false);
    });

    test("should not hide the overlay if any other element is clicked", () => {
        const singleLineElement = document.createElement("p");
        singleLineElement.setAttribute(
            "data-cslp",
            "all_fields.blt58a50b4cebae75c5.en-us.single_line"
        );

        document.body.appendChild(singleLineElement);

        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);
        singleLineElement.click();
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);
    });

    test("should send update field event to the parent if the focused element is an inline editable element", async () => {
        editedElement.setAttribute("contenteditable", "true");

        // We"ll always click one of the overlays, so we can just grab the first one. Manually pointing the global state to the editedElement as we are not simulating mouse click on window here.
        VisualEditor.VisualEditorGlobalState.value.previousSelectedEditableDOM =
            editedElement;

        // already called addFocusOverlay, hence visible is set to true
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);

        await fireEvent.change(editedElement, {
            target: { innerHTML: "New text" },
        });

        expect(editedElement.textContent).toBe("New text");

        // close the overlay
        await fireEvent.click(focusOverlayWrapper);
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(false);

        expect(liveEditorPostMessage?.send).toHaveBeenCalledTimes(1);
        expect(liveEditorPostMessage?.send).toHaveBeenCalledWith(
            LiveEditorPostMessageEvents.UPDATE_FIELD,
            {
                data: editedElement.textContent,
                fieldMetadata: {
                    content_type_uid: "all_fields",
                    cslpValue: "all_fields.blt58a50b4cebae75c5.en-us.title",
                    entry_uid: "blt58a50b4cebae75c5",
                    fieldPath: "title",
                    fieldPathWithIndex: "title",
                    locale: "en-us",
                    multipleFieldMetadata: { index: -1, parentDetails: null },
                },
            }
        );
    });

    test("should run cleanup function", () => {
        // We"ll always click one of the overlays, so we can just grab the first one. Manually pointing the global state to the editedElement as we are not simulating mouse click on window here.
        VisualEditor.VisualEditorGlobalState.value.previousSelectedEditableDOM =
            editedElement;
        fireEvent.click(focusOverlayWrapper);

        expect(cleanIndividualFieldResidual).toHaveBeenCalledTimes(1);
    });

    // TODO
    test("should hide the overlay if the escape key is pressed", () => {
        expect(focusOverlayWrapper.classList.contains("visible")).toBe(true);

        const escapeEvent = new KeyboardEvent("keydown", {
            key: "Escape",
        });
        window.dispatchEvent(escapeEvent);

        expect(focusOverlayWrapper.classList.contains("visible")).toBe(false);
    });
});
