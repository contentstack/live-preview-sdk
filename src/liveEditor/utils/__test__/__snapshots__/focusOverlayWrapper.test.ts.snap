// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addFocusOverlay should set the top overlay to the correct dimensions 1`] = `
<div
  class="visual-editor__overlay__wrapper"
  data-testid="visual-editor__overlay__wrapper"
>
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--top"
    data-testid="visual-editor__overlay--top"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--left"
    data-testid="visual-editor__overlay--left"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--right"
    data-testid="visual-editor__overlay--right"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--bottom"
    data-testid="visual-editor__overlay--bottom"
  />
  
        
  <div
    class="visual-editor__overlay--outline"
    data-testid="visual-editor__overlay--outline"
  />
  
    
</div>
`;

exports[`addFocusOverlay should set the top overlay to the correct dimensions 2`] = `
<div
  class="visual-editor__overlay__wrapper visible"
  data-testid="visual-editor__overlay__wrapper"
>
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--top"
    data-testid="visual-editor__overlay--top"
    style="top: 0px; left: 0px; width: 100%; height: 10px;"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--left"
    data-testid="visual-editor__overlay--left"
    style="left: 0px; top: 10px; width: 10px;"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--right"
    data-testid="visual-editor__overlay--right"
    style="left: 20px; top: 10px; width: 1004px;"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--bottom"
    data-testid="visual-editor__overlay--bottom"
    style="top: 20px; height: -20px; left: 0px; width: 100%;"
  />
  
        
  <div
    class="visual-editor__overlay--outline"
    data-testid="visual-editor__overlay--outline"
    style="top: 10px; left: 10px;"
  />
  
    
</div>
`;
